src/listen/ldap_sync/proto_ldap_sync_ldap.c

```diff
@@ -0,0 +1,1325 @@
+	{ NULL }
+};
+
+/** Operations performed on entries
+ */
+fr_table_num_sorted_t const sync_op_table[] = {
+	{ L("add"),			SYNC_OP_ADD			},
+	{ L("delete"),			SYNC_OP_DELETE			},
+	{ L("modify"),			SYNC_OP_MODIFY			},
+	{ L("present"),			SYNC_OP_PRESENT			},
+};
+size_t sync_op_table_len = NUM_ELEMENTS(sync_op_table);
+
+/** Context used when looking up Directory types
+ */
+typedef struct {
+	fr_listen_t		*main_listen;
+	fr_listen_t		*child_listen;
+	fr_connection_t		*conn;
+	int			msgid;
+} proto_ldap_dir_ctx;
+
+/** Compare two sync state structures on msgid
+ *
+ * @param[in] one first sync to compare.
+ * @param[in] two second sync to compare.
+ * @return CMP(one, two)
+ */
+int8_t sync_state_cmp(void const *one, void const *two)
+{
+	sync_state_t const *a = one, *b = two;
+
+	return CMP(a->msgid, b->msgid);
+}
+
+/** Tell the remote server to stop the sync
+ *
+ * Terminates the search informing the remote server that we no longer want to receive results
+ * for this sync.  A RFC 4511 abandon request is used to inform the server.
+ *
+ * This allows individual syncs to be stopped without destroying the underlying connection.
+ *
+ * Removes the sync's msgid from the tree of msgids associated with the connection.
+ *
+ * @param[in] sync to abandon.
+ * @return 0
+ */
+static int sync_state_free(sync_state_t *sync)
+{
+
+	fr_ldap_connection_t	*conn = talloc_get_type_abort(sync->conn, fr_ldap_connection_t);
+	fr_rb_tree_t		*tree = talloc_get_type_abort(conn->uctx, fr_rb_tree_t);
+	sync_state_t		find = { .msgid = sync->msgid };
+
+	DEBUG3("Abandoning sync");
+
+	if (!sync->conn->handle) return 0;	/* Handled already closed? */
+
+	/*
+	 *	Tell the remote server to stop sending results
+	 */
+	if (sync->msgid >= 0) ldap_abandon_ext(sync->conn->handle, sync->msgid, NULL, NULL);
+	fr_rb_delete(tree, &find);
+
+	return 0;
+}
+
+/** Allocate a sync state
+ *
+ * @param[in] ctx	to allocate the sync state in.
+ * @param[in] conn	which the sync will run on.
+ * @param[in] config	for the sync.
+ * @return new sync state.
+ */
+sync_state_t *sync_state_alloc(TALLOC_CTX *ctx, fr_ldap_connection_t *conn, sync_config_t const *config)
+{
+	sync_state_t	*sync;
+
+	MEM(sync = talloc_zero(ctx, sync_state_t));
+	sync->conn = conn;
+	sync->config = config;
+	sync->phase = SYNC_PHASE_INIT;
+
+	/*
+	 *	If the connection is freed, all the sync state is also freed
+	 */
+	talloc_set_destructor(sync, sync_state_free);
+
+	return sync;
+}
+
+/** Destructor for sync_entry_packet_t
+ *
+ * Ensures LDAP messages are freed
+ */
+static int _sync_entry_packet_free(sync_entry_packet_t *sync_packet)
+{
+	if (sync_packet->msg) ldap_msgfree(sync_packet->msg);
+
+	return 0;
+}
+
+/** Serialise values from sync config for sending to worker.
+ *
+ * @param[out] dbuff	to append attributes to
+ * @param[in] config	to read values from
+ */
+static int serialise_sync_config_values(fr_dbuff_t *dbuff, sync_config_t const *config)
```

Comment:
It'd make more sense here to use the internal encoder and decoder, then the pairs can just pop into existence when they're decoded on the other side?

Reply:
It makes it less work for us to maintain because we only need to update the serialization function

Reply:
I envisaged similar issues with the async redis code, which was the reason for writing the internal encoder/decoder